pkg/querier/blocks_store_queryable_test.go

```diff
@@ -544,6 +545,474 @@ func TestBlocksStoreQuerier_Select(t *testing.T) {
+							Warnings: []string{},
+							Hints:    mockValuesHints(block3),
+						},
+					}: {block3, block4},
+				},
+				// Third attempt returns the last missing block.
+				map[BlocksStoreClient][]ulid.ULID{
+					&storeGatewayClientMock{
+						remoteAddr: "4.4.4.4",
+						mockedLabelNamesResponse: &storepb.LabelNamesResponse{
+							Names:    []string{},
+							Warnings: []string{},
+							Hints:    mockNamesHints(block4),
+						},
+						mockedLabelValuesResponse: &storepb.LabelValuesResponse{
+							Values:   []string{},
+							Warnings: []string{},
+							Hints:    mockValuesHints(block4),
+						},
+					}: {block4},
+				},
+			},
+			expectedLabelNames:  namesFromSeries(series1, series2),
+			expectedLabelValues: valuesFromSeries(labels.MetricName, series1, series2),
+			expectedMetrics: `
+				# HELP cortex_querier_storegateway_instances_hit_per_query Number of store-gateway instances hit for a single query.
+				# TYPE cortex_querier_storegateway_instances_hit_per_query histogram
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="0"} 0
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="1"} 0
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="2"} 0
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="3"} 0
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="4"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="5"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="6"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="7"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="8"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="9"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="10"} 1
+				cortex_querier_storegateway_instances_hit_per_query_bucket{le="+Inf"} 1
+				cortex_querier_storegateway_instances_hit_per_query_sum 4
+				cortex_querier_storegateway_instances_hit_per_query_count 1
+
+				# HELP cortex_querier_storegateway_refetches_per_query Number of re-fetches attempted while querying store-gateway instances due to missing blocks.
+				# TYPE cortex_querier_storegateway_refetches_per_query histogram
+				cortex_querier_storegateway_refetches_per_query_bucket{le="0"} 0
+				cortex_querier_storegateway_refetches_per_query_bucket{le="1"} 0
+				cortex_querier_storegateway_refetches_per_query_bucket{le="2"} 1
+				cortex_querier_storegateway_refetches_per_query_bucket{le="+Inf"} 1
+				cortex_querier_storegateway_refetches_per_query_sum 2
+				cortex_querier_storegateway_refetches_per_query_count 1
+			`,
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Splitting it because we need a new registry for names and values.
+			// And also the initial expectedErr checking needs to be done for both.
+			for _, testFunc := range []string{"LabelNames", "LabelValues"} {
```

Comment:
I feel like there should be a saner approach than this but my brain can't come up with any. Any ideas @pracucci / @pstibrany?

Reply:
I think you did great. I would have probably done the same. We have other tests where we adopt this approach. LGTM.